doc/performance/0002-http-vs-rcp-client-performance-analysis.md

```diff
@@ -0,0 +1,244 @@
+|         | Current                        | JSON-RPC | Current                     | JSON-RPC | Current                     | JSON-RPC | Current             | JSON-RPC | JSON-RPC |
+| Median  | 1346                           | 753      | 0                           | 2        | 0                           | 97       | 1347                | 852      | 99       |
+| Average | 1339                           | 765.3    | 0                           | 2.1      | 0                           | 103.3    | 1339                | 870.7    | 105.4    |
+| Max     | 1713                           | 830      | 1                           | 3        | 1                           | 149      | 1714                | 936      | 152      |
+| Min     | 877                            | 719      | 0                           | 1        | 0                           | 69       | 877                 | 807      | 72       |
+
+## Analysis
+
+### TSPClient processing time
+
+In all examples, the TSPClient processing time of the current implementation is much slower compared to the JSON-RPC implementation. The reason is that, in the case of JSON-RPC, the TSPClient code is ran in NodeJS, instead of the browser. NodeJS uses the v8 compiler, which is a JIT compiler which [optimizes code performance](https://blog.bitsrc.io/the-jit-in-javascript-just-in-time-compiler-798b66e44143). In addition, the Theia BE uses a zero copy serializer for its BE-Trace Server communication. Hence, the TSPClient runs a lot faster on the Theia BE. From the numbers that we have, the speed of the TSPClient improves at least `2 times` and can be up to `9 times` when running in the BE.
+
+### Front-end to TSPClient
+
+In all test cases, the time takes to reach the TSPClient from the Theia FE is pretty similar. In the first 3 test cases (1, 2, and 3a), it is inconclusive to say that there is a significant difference in speed. In test case 3b, by removing other factors such as parallel running HTTP requests, the time takes to reach the TSPClient in the JSON-RPC implementation is only `2ms` more than the current implementation, which is not really significant. However, payloads going to the Theia BE test here is small. Larger parameters might contribute more latency. However, we don't expect the request payload to become extremely large, such as the case of the response payload.
+
+### TSPClient to Front-end
+
+In all test cases, there is virtually no latency for returning the HTTP response to the front-end in the current implementation. However, there is significant latency when using JSON-RPC.Latency scales with the size of the response. The larger the response is, the longer the latency. When getting out descriptors, where we get a small response payload, the latency added is approximately `50ms`. In test case 3b, where we fetch around `116620 states`, latency can be `150ms`.
```

Comment:
There is a space missing after `JSON-RPC.` and before `Latency`

Reply:
Ack.